### ISSUE TITLE
Error Handling in C++ or: Why We Should Use Eithers in Favor of Exceptions and Error-codes

### ISSUE THREAD

**Issue by cogutvalera** [bitshares/bitshares-core#1338]
This issue was born after #946 issue where we have to check if there was caught exception or not by `std::current_exception()` method.

```
   // ************************************************************************************************************** //
   // Issue summary:          There is a method std::current_exception() for checking caught exception, at the same  //
   // time it depends on error handling implementation. Currently is implemented throw/catch approach in which       //
   // are disadvantages like: performance overhead, exceptions aren't supported by all platforms,                    //
   // the biggest drawback is that handling exceptions is not enforced by the type-system.                           //
   // Unlike, Java, for example, where exceptions must be caught by the caller,                                      //
   // catching a C++ exception is optional. This means spotting all the unhandled exceptions during a code review    //
   // will be challenging, and requires deep knowledge of all of the functions called.                               //
   // Consider using an Either type to handle errors. They lift the error into the type-system,                      //
   // making them safer than exceptions whilst yielding the same performance characteristics as error-codes.         //
   // ************************************************************************************************************** //
```

1. https://hackernoon.com/error-handling-in-c-or-why-you-should-use-eithers-in-favor-of-exceptions-and-error-codes-f0640912eb45
2. https://github.com/LoopPerfect/neither (C++ Either implementation)

**Comment by pmconrad** [comment]
Funny enough, the need to explicitly handle exceptions in Java is seen as a disadvantage by many.
Some newer languages that are compiled to Java bytecode have gotten rid of checked exceptions entirely, and Java frameworks often use an exception hierarchy that is based on unchecked exceptions as well.
https://programming.guide/java/checked-exceptions-good-or-bad.html

**Comment by cogutvalera** [comment]
@pmconrad I didn't tell that Java Error Handling approach is better than C++
I just told that `try/catch/throw` approach has disadvantages like: 
1. performance overhead
2. exceptions aren't supported by all platforms
3. the biggest drawback is that handling exceptions is not enforced by the type-system

I just wanted to tell that catching a C++ exception is optional while catching a Java exception is obligated and is more simple for code review than in C++, but Java implementation isn't better.

Main idea of my message was to tell about another approach by `Either` for Error Handling instead of `try/catch/throw`.

Thanks !

**Comment by cogutvalera** [comment]
**"A proposal to add a utility class to represent expected monad in C++"** pdf file about proposal for new `Expected` feature similar to `Either`
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4109.pdf

**Comment by cogutvalera** [comment]
IMHO we should divide this "gigantic" issue to many smaller sub-issues

**Comment by cogutvalera** [comment]
@pmconrad Thanks ! Just trying to figure out the best approach for error handling mechanism. Let's discuss for sure all advantages and disadvantages, maybe we don't need it.

**Comment by jmjatlanta** [comment]
< soapbox >

Exceptions change command flow when the unexpected happens. They were designed to be used when something happens that is NOT expected. When used correctly, they allow the code above to handle the problem, or defer to code even higher.

Having a client send a transaction that is invalid is an expected behavior. Having a peer attempt to push a bad block is an expected behavior (unfortunately). Such things should be handled within the pre-defined control flow of the method / return type / signaling mechanism as part of the application design. This requires forethought at the design stage, but makes results predictable. It also requires the calling code to know all the possible predictable results that could be returned from a method call.

There are some things that are rare and unexpected. We do not want to code each method to plan for running out of memory, having a disk go bad, or a network connection get cut. These are situations that are (possibly) appropriate to "bubble up" to higher layers, and interrupt the "normal" program flow.

I believe that try/catch within bitshares is an abused system. It is causing some tangled messes and unwanted overhead. Fixing it will require an investment in design, and a bigger investment in refactoring. Until then, we will live in the nest we soiled.

Such things do not upset me. I have lived with things like this before. I don't care to get into religious wars over such things. I just hope that this causes some to think before they code, or even engage in useful discussion as to why my view is perhaps incorrect.

< /soapbox >
